public/recorder.js

```diff
@@ -77,151 +274,125 @@ langSelector.addEventListener("change", () => {
   }
 });
 
-ref.addEventListener("timeupdate", function() {
-  const segment = Math.floor(ref.currentTime / segmentDuration);
-  const lang = lyrics[selectedLang] ? selectedLang : "en";
-  const line = lyrics[lang][segment];
-  const nextLine = lyrics[lang][segment + 1] || false;
-  if (!line) {
-    karaoke.textContent = "";
-    return;
-  }
-  const current = document.createElement("span");
-  let next;
-  if (nextLine) {
-    next = document.createElement("p");
-    next.className = "next";
-  }
-  if (line.lang) {
-    current.lang = line.lang;
-    current.textContent = line.string
-  } else {
-    current.textContent = line;
-  }
-  if (nextLine) {
-    if (line.lang) {
-      next.lang = nextLine.lang;
-      next.textContent = nextLine.string
-    } else {
-      next.textContent = nextLine;
-    }
-  }
-  karaoke.innerHTML = "";
-  karaoke.appendChild(current);
-  if (nextLine) {
-    karaoke.appendChild(next);
-  }
-});
-
-async function wait(s) {
-  return new Promise(res => setTimeout(res, 1000*s));
-}
+recordBtn.addEventListener("click", async () => {
+  recordBtn.disabled = true;
+  playBtn.disabled = true;
+  pauseBtn.disabled = true;
+  stopLabel.textContent = "Cancel recording";
 
-function resetReplayBtn() {
-  replayBtn.textContent = "Play your last recording";
-  if (recordedAudio) {
-    recordedAudio.pause();
-    recordedAudio = null;
+  if (playbackNode) {
+    await audioContext.suspend();
+    resetAudioGraph();
   }
-}
 
-function stopRecording({ cancel } = { cancel: false }) {
-  recordingCanceled = cancel;
-  ref.pause();
-  stream.getAudioTracks()[0].stop();
-  recorder.stop();
-}
+  // Disable upload form
+  [...form.querySelectorAll("input,button,select")].forEach(n => n.disabled = true);
+  document.getElementById("uploader").classList.add("disabled");
 
-recordBtn.addEventListener("click", async () => {
-  if (onAir) {
-    onAir = false;
-    stopRecording({ cancel: true });
+  // Safari compat notes:
+  // - The following two lines look pretty useless. They are not when recording
+  // for the first time! Safari apparently refuses to set the internal
+  // [[suspended by user]] flag  unless "resume" has been called first... and
+  // "resume" can only be called in response to user activation (so not when
+  // the page loads). Without these lines, the audio context woult start playing
+  // on Safari as soon as a node gets created in the audio graph, not when
+  // "resume" is called.
```

Comment:
out of curiosity, do you know if it is a safari bug (and if so, has it been filed?), or is it allowed by the spec and a design choice?

Reply:
I believe that is a bug. Safari does not reject the call to `suspend()` when the page loads, and according to the spec, that should set the `[[suspended by user]]` internal flag, which should in turn prevent the context from playing immediately when a node gets added:
https://webaudio.github.io/web-audio-api/#dom-audiocontext-suspend

I do not know whether the bug has been filed

Reply:
I prepared a test file and filed a bug against Webkit:
https://bugs.webkit.org/show_bug.cgi?id=216972

... noting that it may actually be worth updating the spec as well, which led me to:
https://github.com/WebAudio/web-audio-api/issues/2251

Reply:
You rulez!